pkg/operator/controller/canary/controller.go

```diff
@@ -306,17 +311,48 @@ func (r *reconciler) startCanaryRoutePolling(stop <-chan struct{}) error {
 	return nil
 }
 
-func (r *reconciler) setCanaryFailingStatusCondition() error {
+func (r *reconciler) setCanaryFailingStatusCondition(errors []error) error {
+	errorStrings := DeduplicateErrorStrings(errors)
+	if len(errorStrings) > 3 {
+		errorStrings = errorStrings[len(errorStrings)-3:]
+	}
 	cond := operatorv1.OperatorCondition{
 		Type:    ingresscontroller.IngressControllerCanaryCheckSuccessConditionType,
 		Status:  operatorv1.ConditionFalse,
 		Reason:  "CanaryChecksRepetitiveFailures",
-		Message: "Canary route checks for the default ingress controller are failing",
+		Message: fmt.Sprintf("Canary route checks for the default ingress controller are failing. Last %d error messages:\n%s", len(errorStrings), strings.Join(errorStrings, "\n")),
 	}
 
 	return r.setCanaryStatusCondition(cond)
 }
 
+// DeduplicateErrorStrings takes a slice of errors in chronological order, prunes the list for unique error messages
+// It returns a slice of the error message as strings, with any string seen more than once incuding how many times it
+// was seen
+// The chronological order is preserved, based on the last time each error message was seen
+func DeduplicateErrorStrings(errors []error) []string {
+	encountered := map[string]int{}
+	ret := []string{}
+	// Iterate over the list of errors from newest to oldest, keeping track of how many times each error message was
+	// seen
+	for i := len(errors) - 1; i >= 0; i-- {
+		if encountered[errors[i].Error()] >= 1 {
+			encountered[errors[i].Error()] += 1
+			continue
+		}
+		encountered[errors[i].Error()] = 1
+		ret = append([]string{errors[i].Error()}, ret...)
```

Comment:
I had to iterate over the input list backwards to keep the most recent errors at the end of the list, but I'm not sure which is better: (a) building the list by "pushing" new items onto the front of the list (the way it's implemented here), or (b) building it from newest to oldest, then reverse it just before returning.

Reply:
If we are talking efficiency, isn't `a` slightly more efficient? I think `a` is `O(n)` where as `b` is `O(2n)`, though `O(2n) == O(n)` as it scales to infinity, so it really doesn't matter a whole lot. Just trying to practice my O Notation skills :smiley:

Reply:
It depends I think on how append actually works in this case. `append(slice, newItem)` is usually O(1), but the `ret...`syntax here is more like `append(slice, ret[0], ret[1], ... , ret[n])`. If it's unoptimized, that could be an O(n) operation, and since that's potentially happening for every iteration of the for loop, this would be O(n<sup>2</sup>)

Reply:
Yea I see that now. I don't completely understand the mechanics of `append` and how it's optimized, but sounds like `b` would be more efficient, being `O(n)` for the building of the unordered slice, then another `O(n)` to reverse, so `2O(n) == O(n)`.